hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+			// and some child resources reuse the same "Properties" type, which could false some of the logic below
+			return astmodel.IdentityVisitOfObjectType(this, it, ctx)
+		}
+
+		// Before visiting, check if any properties are just referring to one of our sub-resources and remove them
+		subResources := e.resourceToSubresourceMap[typedCtx.resource]
+		for _, prop := range it.Properties() {
+			propTypeName, ok := astmodel.AsTypeName(prop.PropertyType())
+			if !ok {
+				continue
+			}
+
+			// TODO: This is currently no different than the below, but it likely will evolve to be different over time
+			if subResources.Contains(propTypeName) {
+				klog.V(5).Infof("Removing resource %q reference to subresource %q on property %q", typedCtx.resource, propTypeName, prop.PropertyName())
+				it = removeResourceLikeProperties(it)
+				continue
+			}
+
+			if e.resourcePropertiesTypes.Contains(propTypeName) {
+				klog.V(5).Infof("Removing reference to resource %q on property %q", propTypeName, prop.PropertyName())
+				it = removeResourceLikeProperties(it)
+			}
+		}
+
+		return astmodel.IdentityVisitOfObjectType(this, it, ctx)
+	}
+
+	return visitor
+}
+
+func (e embeddedResourceRemover) NewResourceRemovalTypeWalker(visitor astmodel.TypeVisitor) *astmodel.TypeWalker {
+	typeWalker := astmodel.NewTypeWalker(e.types, visitor)
+	typeWalker.AfterVisit = func(original astmodel.TypeDefinition, updated astmodel.TypeDefinition, ctx interface{}) (astmodel.TypeDefinition, error) {
+		typedCtx := ctx.(resourceRemovalVisitorContext)
+
+		if !original.Name().Equals(updated.Name()) {
+			panic(fmt.Sprintf("Unexpected name mismatch during type walk: %q -> %q", original.Name(), updated.Name()))
+		}
+
+		if !original.Type().Equals(updated.Type()) {
+			flaggedType := e.typeFlag.ApplyTo(updated.Type())
+			var newName astmodel.TypeName
+			exists := false
+			for count := 0; ; count++ {
+				newName = embeddedResourceTypeName{original: original.Name(), context: typedCtx.resource.Name(), suffix: e.typeSuffix, count: count}.ToTypeName()
+				existing, ok := typedCtx.modifiedTypes[newName]
+				if !ok {
+					break
+				}
+				if existing.Type().Equals(flaggedType) {
+					exists = true
+					// Shape matches what we have already, can proceed
+					break
+				}
+			}
+			updated = updated.WithName(newName)
+			updated = updated.WithType(flaggedType)
+			if !exists {
+				typedCtx.modifiedTypes.Add(updated)
+			}
+
+			klog.V(5).Infof("Updating %q to %q", original.Name(), updated.Name())
+		}
+
+		return updated, nil
+	}
+
+	typeWalker.ShouldRemoveCycle = func(def astmodel.TypeDefinition, ctx interface{}) (bool, error) {
+		// If we're about to walk a cycle that is to a known resource type, just skip it entirely
+		if e.resourcePropertiesTypes.Contains(def.Name()) || e.resourceStatusTypes.Contains(def.Name()) {
+			return true, nil
+		}
+
+		if isTypeResourceLookalike(def.Type()) {
```

Comment:
Added a comment explaining why this is here.

Reproducing below:
```
// This is here because some microsoft.networking resources are resources (in the sense that they have ARM IDs)
// but can only be created as children of another resource. The resources in question don't have
// their own PUT and so are not actually classified as a top level resource by the JSON schema. We don't want to
// remove ALL cycles in the type graph currently as we can't know for sure that the cycles are structurally meaningless.
// This is an attempt at a middle-ground heuristic that lets us find cycles between things that are resource-like.
// For example see the cycle between NetworkInterfaceIPConfiguration in microsoft.network 20180601:
// NetworkInterfaceIPConfiguration_Status -> NetworkInterfaceIPConfigurationPropertiesFormat_Status ->
// ApplicationGatewayBackendAddressPool_Status -> ApplicationGatewayBackendAddressPoolPropertiesFormat_Status -> NetworkInterfaceIPConfiguration_Status
// Sometimes these resource-like things are promoted to real resources in future APIs as in the case of Subnet in the 2017-06-01
// API version.
```